rfcs/rfc-authentication-partner-applications.md

```diff
@@ -0,0 +1,89 @@
+
+Intention of this RFC is to list down some potential solution options and requests for comments or alternate solutions.
+
+
+# Solution Options
+## 1) Client application to Integrate with Keycloak Integration
+### Summary
+Client redirects to keycloak where the user authenticates and gets redirected back to their application on succesfull login. Client application would need to do the code exchange to get the access and refresh tokens. 
+
+![Keycloak integration from partner application](rfc-authentication-partner-applications/KeyCloak.png)
+
+
+#### Pros
+- Less work on registries side, would be mostly configuration
+
+#### Cons
+- This approach solves only authentication; user wouldn't be able to setup their account or registries won't be able to perform any validations/business rules before the user is sent to client application.
+- Introduces KeyCloak dependency on client application. Client application would need to keep the keycloak configs and would need to handle all integration with keycloak from their application. 
+
+## 2) Perform a document post with tokens
+### Summary
+Upon authentication for the user to navigate to the client application, perform a document POST with tokens populated as elements. 
+
+![Document Post](rfc-authentication-partner-applications/Post.png)
+
+
+#### Pros
+- Client application doesn't have Keycloak dependency as in option 1
+#### Cons
+- This will work only if the client application is using server side rendering.
+
+
+## 3) Cross-Domain Messaging using Iframe
+### Summary
+Registries application would post an iframe message to document in client application with tokens, and client application can receive the tokens using window event listener.
+![IFrame Post Message](rfc-authentication-partner-applications/IFrame.png)
+
+#### Pros
+- Easy to implement and has been initially adopted in BC Registries before moved all apps under same domain
+
+
+#### Cons
+- Registries app would need to load the client application in iFrame.
+- If the client application has X-FRAME-OPTIONS to DENY or SAMEORIGIN, this will prevent the app being loaded in iFrame and won't be able to receive the tokens. For web applications this value is normally not provided and allows the app to be included as IFrame unless specified, but some portal platform enforces this.
+
+## 4) Cross-domain Cookie
+### Summary
+Registries application would set cookies in client application domain. Client applications need to create a dynamic page where it sets the cookie. Registries application creates a hidden image upon user authentication which points to the dynamic page on the client application, and pass the tokens. Registries team would be able to provide a javascript util to handle this. 
+
+![Cookie](rfc-authentication-partner-applications/Cookie.png)
+#### Pros
+- Similar to iFrame approach and has been generally by some known products. 
+
+#### Cons
+- If third party cookie is disabled in browser, the solution won't work.
+
```

Comment:
FYI there's some new Chrome restriction on cross-site cookies coming in Feb. As far as I understand this is not a blocker, but will require the specific cross-domain setting (SameSite) to work. Might be a problem in unsupported browsers. https://blog.chromium.org/2019/10/developers-get-ready-for-new.html

Reply:
Agreed. I don't personally prefer this solution as it depends on browser features. :)